src/Mod/Draft/draftobjects/patharray.py

```diff
@@ -119,7 +119,7 @@ def setProperties(self,obj):
             obj.addProperty("App::PropertyVector","VerticalVector","Alignment", _tip)
             obj.VerticalVector = App.Vector(0,0,1)
 
-        if self.use_link:
+        if hasattr(obj, "use_link"):
```

Comment:
Are you sure the the `use_link` isn't added automatically? It should be added in the `onDocumentRestored` method, but only for `PathLinkArrays`, not for regular arrays. Do you have the test file to investigate?

I think your code won't work, because `obj.use_link` shouldn't exist. What exists is `obj.Proxy.use_link`. This `use_link` attribute is an attribute of the Proxy class not of the object. The proper test should be `hasattr(self, "use_link")`.

Actually this part needs a test for `ExpandArray`, otherwise this property is added again every time a PathLinkArray is loaded.
```py
if self.use_link and not hasattr(obj, "ExpandArray"):
```

I think the actual problem is in `onDocumentRestored`. The migration of the properties should be done before setting the properties.
```py
def onDocumentRestored(self, obj):
    self.migrate_attributes(obj)
    self.set_properties(obj)
```

I have a branch where I'm cleaning a bit this stuff. So, it would be good if you can pass me an old file that doesn't work to investigate.

The link parameter was introduced with the LinkMerge, and then it was slightly altered by me a few months later, so faulty files should exist only for a period of a few months from Augusto 2019 to January 2020 maybe.

Reply:
Discussion cont's at https://forum.freecadweb.org/viewtopic.php?f=23&t=46176&p=402523#p402523 where there is a link for the test file

Reply:
I just tested my patch, and it seems to work. As I said, just use the migrate function before setting the properties.
```py
    def onDocumentRestored(self, obj):
        self.migrate_attributes(obj)
        self.setProperties(obj)
```

The rest of the code doesn't need to be touched at the moment, but if you want you can also add this.
```py
        if self.use_link and "ExpandArray" not in pl:
            _tip = _tr("Show array element as children object")
            obj.addProperty("App::PropertyBool","ExpandArray", "Parameters", _tip)
            obj.ExpandArray = False
            obj.setPropertyStatus('Shape','Transient')
```